tests/test_global_opts.rs

```diff
@@ -47,23 +47,11 @@ fn test_non_utf8_arg() {
 #[test]
 fn test_no_subcommand() {
     let test_env = TestEnvironment::default();
+    test_env.jj_cmd_success(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
 
-    let stderr = test_env.jj_cmd_cli_error(test_env.env_root(), &[]);
-    insta::assert_snapshot!(stderr.lines().next().unwrap(), @"Jujutsu (An experimental VCS)");
-
-    let stderr = test_env.jj_cmd_cli_error(test_env.env_root(), &["-R."]);
-    insta::assert_snapshot!(stderr.lines().next().unwrap(), @"error: 'jj' requires a subcommand but one was not provided");
-
-    let stdout = test_env.jj_cmd_success(test_env.env_root(), &["--version"]);
-    let sanitized = stdout.replace(|c: char| c.is_ascii_hexdigit(), "?");
-    assert!(
-        sanitized == "jj ?.?.?\n"
-            || sanitized == "jj ?.?.?-????????????????????????????????????????\n",
-        "{sanitized}"
-    );
-
-    let stdout = test_env.jj_cmd_success(test_env.env_root(), &["--help"]);
-    insta::assert_snapshot!(stdout.lines().next().unwrap(), @"Jujutsu (An experimental VCS)");
+    let stdout = test_env.jj_cmd_success(&repo_path, &[]);
+    assert_eq!(stdout, test_env.jj_cmd_success(&repo_path, &["log"]));
 }
```

Comment:
I think we should not change the behavior of `jj --version` and `jj --help`. We should probably allow `jj -R <path>` and make that show the log. Plain `jj` outside of a repo should probably still show help. That seems to match what `sl` does.

Do we want the default command to be configurable, maybe even including flags?

I should say that I'm still a bit hesitant about this feature overall, mostly because I find it a bit strange that `sl` behaves like `sl sl` but `sl -r .` does not behave like `sl sl -r .`. Actually, why doesn't it? It seems like it should work to make it accept flags for the implied subcommand.

So, overall, maybe a better way to implement this is a to it similar to [alias resolution](https://github.com/martinvonz/jj/blob/cf402af9c15755064cc3cc435b2c02223e51354d/src/cli_util.rs#L2031-L2095), maybe just before we do that, i.e. just before https://github.com/martinvonz/jj/blob/cf402af9c15755064cc3cc435b2c02223e51354d/src/cli_util.rs#L2135.

What do you think?

Reply:
Since `--version` and `--help` and `-R` are global options, those will still work as expected. I'll add those tests back in.

I wasn't able to figure out how to get `jj` to show help when outside of a repo, but I'll poke at it a bit more. It may just take a conditional when `None` matches, but I'll see if that information is in scope.

Not being able to parse the implied subcommand flags was a limitation I wasn't able to figure out, and I think that's partially why you're also seeing some inconsistent behavior with `sl`. I can iterate a bit with the configurable idea and/or using something similar to alias resolution to see if can be made more consistent.

Reply:
Thanks.

Defaulting to `help` when outside of a repo is actually going to be harder if we treat it like alias resolution. That's probably fine (getting `Error: There is no jj repo in "."` seems reasonable).

Reply:
That's a bit surprising given that you removed `.subcommand_required(true)`, but I haven't thought much about it.

> I haven't had time yet to try out your strategy of resolving the command earlier. That'll be next.

Sounds good. I think it will be much easier to handle arguments that way (and to make the default command configurable, if we ever want to do that).